### PR TITLE
Convert `enarx-pr-merge` to GraphQL, and make it consumable by other repos

### PR DESCRIPTION
**Pull request by mbestavros**
Converts the auto-merge bot action to use pure GraphQL, and to run as a scheduled action on implementing repos.

This is purely additive; the existing bot functionality should continue to work. If used, this action would simply run all the PR-related scripts when a pull request target event happens.

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+description: Github Actions project automation for Enarx
+branding:
+  icon: git-commit
+  color: green
+
+inputs:
+  GITHUB_TOKEN:
+    description: "An appropriately-scoped token to authenticate with Github."
+    required: true
+  EVENT_NAME:
```
Comment (npmccallum) [inline action.yml:11]:
```suggestion
  GITHUB_EVENT_NAME:
```

File: action.yml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    description: "The event that triggered this run."
+    required: true
```
Comment (npmccallum) [inline action.yml:13]: We probably want a variety of other GH environment variables too. At least:
* `GITHUB_ACTOR`
* `GITHUB_REPOSITORY`
* `GITHUB_SHA`
* `GITHUB_REF`
* `GITHUB_HEAD_REF` (optional)
* `GITHUB_BASE_REF` (optional)
* `GITHUB_API_URL`
* `GITHUB_GRAPHQL_URL`

We also probably want the contents of the file at `GITHUB_EVENT_PATH`.

We won't use these variables yet. But it will be much easier to ignore variables we don't need than to fix every repo that uses these actions later.

File: entrypoint [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/python3
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+
+event = os.environ.get('EVENT_NAME', None)
+if event == "pull_request_target":
+    os.system("enarx-pr-request")
+    os.system("enarx-pr-merge")
+    os.system("enarx-pr-assign")
```
Comment (npmccallum) [inline entrypoint:10]: Doing this will very likely cause us to exceed the GH API rate limiting. It would be better to port one single action to efficiently use the input data.

File: enarx-pr-request-event [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+path = os.environ.get('GITHUB_EVENT_PAYLOAD', None)
+event = json.load(path)
+
+# Get reviews team
+org = github.get_organization("enarx")
+reviews = {t.name: t for t in org.get_teams()}['reviews']
+potential_reviewers = {r.login for r in reviews.get_members()}
+
+# Auto-assign for all PRs in the org.
+pr_number = event["pull_request_target"]["number"]
```
Comment (mbestavros) [inline enarx-pr-request-event:45]: I don't have access to a good reference for what the `GITHUB_EVENT` payload looks like, so this is an educated guess. I'll need to do some testing when possible in order to figure out if this is correct.

File: enarx-pr-merge-event [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import enarxbot
+import sys
+
+# Initialize Github
+github = enarxbot.connect()
+
+# Take inputs from shell and grab the PR to work on.
+(repo_name, number) = sys.argv[1:3]
+repo = github.get_repo(repo_name)
+pr = repo.get_pull(number)
```
Comment (npmccallum) [inline enarx-pr-merge-event:13]: These are going to cause network access. But we only need access to `mergeable` and `mergeable_state`, both of which are already contained in `GITHUB_EVENT_PAYLOAD`. That is, we have all the information we need to make the decision to merge or not already. We need to make only a single network call to perform the actual merge.

So let's ditch PyGithub and just do a graphql mutation for the merge.

File: enarx_event_pr_merge.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import enarxbot
+
+MERGEABLE_QUERY = '''
+query($owner:String!, $repo:String!) {
+  repository(owner: $owner, name: $repo) {
+    pullRequests(states: OPEN, first: 100) {
```
Comment (npmccallum) [inline enarx_event_pr_merge.py:8]: Depaginate this...

**Comment by haraldh** [comment]
I am sorry, I have no idea, if this is correct.

File: enarxbot-pr-merge [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import bot
+
+import json
+import sys
+import os
+
+MERGEABLE_QUERY = '''
+query($owner:String!, $repo:String!, $cursor:String) {
```
Comment (lkatalin) [inline enarxbot-pr-merge:11]: What is a `$cursor`?

**Review by lkatalin — DISMISSED** [review]
I think this looks reasonable, although like with a few other PRs I don't have the full context. What is the difference between the `bot` and the `enarxbot`? Is one of them for GraphQL?

**Comment by mbestavros** [comment]
@lkatalin Sorry for the late response, but yes: we currently have two approaches here.

The first is our existing bot, which has some helper functions within `enarxbot.py`. This is entirely schedule-based and uses Github's REST API via a third-party library (PyGithub). Every hour, each individual action consumes relevant organization state and uses that to determine what actions to take (for example, merging a PR).

We've begun transitioning to a new approach, which calls Github's newer v4 GraphQL API directly instead of its REST API, and also runs on event triggers instead of a schedule. This nets us a few benefits:

- We no longer rely on a third party library to interact with Github. This means all our automation can be run on bare-metal Github runners without a container, so running individual actions becomes very fast.
- Github's GraphQL API seems to expose more data than the REST API, which can be helpful sometimes.
- If applicable, we can also run immediately upon relevant events, which speeds things up even further. For example, our triage action could be run immediately upon somebody opening a new issue instead of waiting for the next turn of the hour, as it currently does.

This PR rewrites (additively) an existing action from approach 1 into approach 2.

File: enarxbot-pr-merge (hunk 7 above)
Comment (mbestavros) [reply enarxbot-pr-merge:11]: @lkatalin See my response to your top-level question for more background on GraphQL.

Since we're calling GraphQL directly now, a big thing we need to deal with is **API pagination**. Asking for all issues in a repository, for example, may be feasible for a small repo, but for a large one it may constitute hundreds of thousands of things. Returning all of them in one JSON object just isn't feasible in some cases.

The solution to this (from an API perspective) is to paginate these results. So, for example, if I ask for issues from a project, Github will return only the first 100. If I want the next 100, I need to "turn the page" and ask for them in a separate request.

This is necessary for Github, but it's quite annoying for us as consumers of the API that just want a clean data structure of everything we ask for. Our solution is to 

1. request pages of objects until there are no more pages
2. stitch these pages together locally to return a unified, usable set for consumers

Basically, whenever you ask Github for paginated data, and properly specify a few arguments (of which `$cursor` is one), our `graphql()` function will automatically "depaginate" the data.

You can see the code that does this here, along with instructions for how to depaginate: https://github.com/enarx/bot/blob/master/bot.py#L72

**Comment by lkatalin** [comment]
Thanks for the detailed explanations, @mbestavros !

**Review by haraldh — APPROVED** [review]

**Review by npmccallum — CHANGES_REQUESTED** [review]
1. Line length.
2. Clean up owner/repo (as discussed).
3. Print before performing the merge action.

File: enarxbot-pr-merge [hunk 8]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/python3
```
Comment (connorkuehl) [inline enarxbot-pr-merge:1]: nit picking:

```suggestion
#!/usr/bin/env python3
```

File: enarxbot-pr-merge [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    # Status strings.
+    merged_indicator = "✗"
+    pr_identifier = f"{owner}/{repo}#{pr['number']}"
+
+    # If all merge criteria pass, PR should be ready to merge.
+    if pr["mergeable"] == "MERGEABLE" and pr["mergeStateStatus"] == "CLEAN":
+        merged_indicator = "✓"
+
+    # List the status of the pull request before attempting to merge.
+    print(f" {merged_indicator:19}| {pr_identifier:23}| {str(pr['mergeable']):11}| {pr['mergeStateStatus']:10}")
```
Comment (connorkuehl) [inline enarxbot-pr-merge:57]: Is the space at the beginning of this string intentional?

**Review by connorkuehl — APPROVED** [review]
LGTM

File: enarxbot-pr-merge (hunk 9 above)
Comment (mbestavros) [reply enarxbot-pr-merge:57]: Yep. Makes the printed results look a bit nicer.

**Review by npmccallum — APPROVED** [review]
